feat: Migrate exercise to issue-based flow

wolz-CODElife opened the pull request:
### Summary
This PR enhances the `connect-the-dots` skills exercise by converting it to an Issue-based flow, ensuring dynamic issue references, and improving workflow reliability and modularity. The changes support a scalable learner experience with better debugging and progression tracking.

### Changes
- [x] Set write permissions for `contents`, `actions`, and `issues` to enable workflow operations.
- [x] Defined the template files(`STEP_1_FILE`, `STEP_2_FILE`, ...,`STEP_n_FILE`) as environment variable for consistent template referencing.
- [x] Implemented `start_exercise` job using `skills/exercise-toolkit/start-exercise.yml@v0.3.0` to initialize the exercise and output the issue URL.
- [x] Added `post_next_step_content` job to:
  - Create dynamic issues ("Fix the broken sidebar" and "Sidebar bug again") using `actions/github-script@v7`.
  - Prepare file history with `_sidebar.md` updates and commit SHA storage.
  - Build and post content with `action-text-variables@v1`, using `{{original_issue}}` and `{{duplicate_issue}}` for dynamic referencing.
  - [x] Made issue identifiers `{{original_issue}}` and `{{duplicate_issue}}` dynamic, replacing hardcoded values to support multiple concurrent issue creations.
  - Change the navigation system to use: `gh workflow enable`
- [x] Added `find_exercise` job to locate the exercise issue using `skills/exercise-toolkit`.
- [x] Added `check_step_work` and `verify_step_work` jobs  to:
  - Post "checking work" feedback and verify learner progress.
  - Use `actions/github-script@v7` for context debugging and output tracking.
- [x] Enhanced `post_next_step_content` with dynamic content and workflow toggling.

<!-- If there's an existing issue for your change, please link to it below next to "Closes".
If there's _not_ an existing issue, please open one first to make it more likely that this update will be accepted. -->

Closes: #94 

### Task list
- [x] For workflow changes, I have verified the Actions workflows function as expected.
- [x] For content changes, I have reviewed the [style guide](https://github.com/github/docs/blob/main/contributing/content-style-guide.md).

FidelusAleksander commented on .github/workflows/0-welcome.yml:
```diff
@@ -1,77 +1,75 @@
+          path: exercise-toolkit
+          ref: v0.3.0
+
+      - name: Create 'Fix broken sidebar' issue (target for duplicate)
```
Create the issues outside of the `post_next_step_content` job

You can have a separate `create_issues` job that is run before `post_next_step_content`

FidelusAleksander commented on .github/workflows/0-welcome.yml:
```diff
@@ -1,77 +1,75 @@
+    uses: skills/exercise-toolkit/.github/workflows/start-exercise.yml@v0.3.0
+    with:
+      exercise-title: "Connect the Dots"
+      intro-message: "Learn to navigate a GitHub repository effectively."
```
Slightly more engaging

```suggestion
      intro-message: "Let's learn to navigate a GitHub repository effectively!"
```

FidelusAleksander commented on .github/steps/1-resolve-duplicate-issues.md:
```diff
@@ -1,4 +1,4 @@
-<!--
+_<!--
```
Remove the underscore, also you can remove all these comments

<img width="533" alt="image" src="https://github.com/user-attachments/assets/c7da837b-6cb8-4deb-8dcf-d955564aaad3" />

FidelusAleksander commented on .github/steps/1-resolve-duplicate-issues.md:
```diff
@@ -24,6 +24,7 @@ When you want to create a crosslink, start typing the title of an issue or pull
-1. Navigate to the issue #1 (Welcome)
-2. Type "Duplicate of #2" as a comment and close issue #1
+1. Navigate to the issue #{{duplicate_issue}} (Welcome)
+2. Type "Duplicate of #{{original_issue}}" as a comment and close issue #{{duplicate_issue}}
```
The original issue probably shouldn't be rendered?

Users won't know they have to write `#2` because the name is rendered

<img width="742" alt="image" src="https://github.com/user-attachments/assets/7e4ae2dc-d06f-4b41-a533-ad38a0378e84" />

FidelusAleksander replied:
Update the step below about "refreshing the page" - this is no longer needed in Issue Based flow.

You can take examples from our other exercises, e.g
`As you close the issue Mona will prepare the next step in this exercise!`

wolz-CODElife replied:
When the learner copies the rendered issue, the URL is copied, so it still appears as `#2`.
And the workflow only Searches for "Duplicate of" || "Duplicate" || "#"

FidelusAleksander commented on .github/workflows/2-find-commit-in-history.yml:
```diff
@@ -1,25 +1,67 @@
     name: Check current step number
     runs-on: ubuntu-latest
+    outputs:
+      current_step: ${{ steps.get_step.outputs.current_step }}
```
What is this for?

wolz-CODElife replied:
I wanted to update the step file, but the `./.github/steps/-step.txt` file is no longer needed in the workflow, so I am removing it.

FidelusAleksander commented on .github/workflows/2-find-commit-in-history.yml:
```diff
@@ -72,10 +105,41 @@ jobs:
+
+      - name: Disable current workflow and enable next one
+        run: |
+          gh workflow disable "Step 2, Find commit in history"
```
Rename the workflows to be simply "Step 2" and "Step 3"

FidelusAleksander commented on .github/workflows/3-fix-broken-sidebar.yml:
```diff
@@ -1,56 +1,62 @@
-name: Step 3, Fix a broken sidebar
+name: Step 3, Fix broken sidebar
```
Rename the workflows to be simply "Step 2" and "Step 3"

You can put the rest after a comment "Step 2 # Fix broken sidebar" but we don't want to have the full thing as a name.

FidelusAleksander reviewed (changes requested):
Please look again at the workflows in `exercise-template` repository, especially what goes into `check_step_work` and `post_next_step_content` 

This documentation section may be helpful for `check_step_work`

https://github.com/skills/skills-manager/blob/main/docs/2-design/design.md#step-grading-and-transition

wolz-CODElife commented:
@FidelusAleksander, I have implemented your suggestions and given it another test run. 
Everything works fine:


<img width="1243" alt="image" src="https://github.com/user-attachments/assets/1292691e-8b27-4bb2-aa48-09bc45487b78" />

![image](https://github.com/user-attachments/assets/2ba0532e-f98e-465b-907d-af5d7ca92f67)

chriswblake commented on .github/steps/x-review.md:
```diff
@@ -21,7 +16,7 @@ You can enable GitHub Pages and see `docs/index.html` as a website!
 1. Click the **Save** button.
 1. Wait about 30 seconds then refresh the page. When you see "Your site is published at ...," you can click on the link to see your published site.
-
+full_repo_name
```
extra text?

chriswblake commented on .github/workflows/0-start-exercise.yml:
```diff
@@ -0,0 +1,153 @@
+      exercise-title: "Connect the Dots"
+      intro-message: "Let's learn to navigate a GitHub repository effectively!"
+
+  check_step_work:
```
I would rename this job to step to better reflect what you are doing. Looks like you are modifying the repository content to prepare the exercise.

chriswblake replied:
The steps below look a bit complex. If it is slow, you might consider running this as a separate job so the onboarding experience is still fast.

chriswblake commented on .github/workflows/0-start-exercise.yml:
```diff
@@ -0,0 +1,153 @@
+      - name: Checkout
+        uses: actions/checkout@v4
+
+      - name: Get response templates
```
Looks like you don't need this.

chriswblake commented on .github/workflows/1-resolve-duplicate-issues.yml:
```diff
@@ -1,78 +1,121 @@
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+
+  verify_step_work:
```
Looks like you created an extra job to split up your content for checking the step. Please look at an existing exercise for an example. There should only be 3 jobs.

chriswblake commented on .github/workflows/1-resolve-duplicate-issues.yml:
```diff
@@ -1,78 +1,121 @@
         uses: actions/checkout@v4
+
+
+      - name: Check for keywords in issue comment
```
I believe the latest version of the toolkit provides a pre-built action for checking if a text body has required keywords.

chriswblake commented on .github/workflows/2-find-commit-in-history.yml:
```diff
@@ -1,53 +1,74 @@
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+
   # Get the current step to only run the main job when the learner is on the same step.
   get_current_step:
```
Please consolidate this step's content into the `check_step_work` job. You can look at an existing exercise for an example.

Docs
https://github.com/skills/skills-manager/blob/main/docs/2-design/design.md#step-grading-and-transition

Example workflow file
https://github.com/skills/github-pages/blob/main/.github/workflows/3-step.yml

chriswblake commented on .github/workflows/2-find-commit-in-history.yml:
```diff
@@ -1,53 +1,74 @@
-      commit_id: ${{ steps.get_commit_id.outputs.commit_id }}
+          echo "commit_id=$(cat .github/files/SIDEBARCOMMIT)" >> $GITHUB_OUTPUT
 
   on_fix_the_sidebar_issue_comment:
```
Please consolidate into the `check_step_work` job.

chriswblake commented on .github/workflows/3-fix-broken-sidebar.yml:
```diff
@@ -1,57 +1,51 @@
+          path: exercise-toolkit
+          ref: v0.3.0
+      
       - name: Dump GitHub pull request context
```
Dev code like should be removed (even if they accidentally left it in the original version).

chriswblake commented on .github/workflows/3-fix-broken-sidebar.yml:
```diff
@@ -61,11 +55,52 @@ jobs:
         run: echo "Found _Closes_ message with autolinked issue in the body of PR $NUMBER"
```
Please see the previously shared example for communicating results back to the user via a results table. The learner should not need to explore the Actions log to find feedback. From their perspective, it just seems like they are interacting with Mona in the issue comments.

chriswblake commented on .github/workflows/3-fix-broken-sidebar.yml:
```diff
@@ -61,11 +55,52 @@ jobs:
+    needs: [find_exercise, check_step_work]
+    runs-on: ubuntu-latest
+    if: |
+      !github.event.repository.is_template
```
FYI, you can remove this in most places. Just keep note of the trigger.

chriswblake commented on .github/steps/1-resolve-duplicate-issues.md:
```diff
@@ -1,11 +1,3 @@
-
 ## Step 1: Resolve duplicate issues
 
 _Welcome to the course :tada:_
```
While you are doing your test run, you will probably notice that the original exercise's "congratulations" type messages are duplicates since Mona is now doing that in the issue comments. You might want to remove these.

chriswblake commented on .github/steps/2-find-commit-in-history.md:
```diff
@@ -30,4 +24,4 @@ A SHA is a reference to a specific object. In this case, it's a reference to a c
 6. Copy the first seven characters of the SHA (the first 7 characters of the 40 character hexadecimal string listed after `commit`)
 7. Comment on issue #2 by adding the SHA from step 6 as a comment text and click on "Comment" button
-8. Wait about 20 seconds then refresh this page (the one you're following instructions from). [GitHub Actions](https://docs.github.com/en/actions) will automatically update to the next step.
+8. As you comment on the issue, Mona will prepare the next step in this exercise!
```
I would reword to 'After you comment...'

chriswblake commented on .github/steps/3-fix-broken-sidebar.md:
```diff
@@ -27,7 +21,7 @@ When you're looking at a commit on GitHub, you can see a lot of information. Fro
+1. In the main branch Edit the [`/docs/_sidebar.md`](../blob/main/docs/_sidebar.md) file.
 2. Correct the spelling of the reference `(doc-references__.md)` on line 4 by changing it into `(doc-references.md)`.
 3. Select or create a new branch `fix-sidebar` for this commit and start a pull request.
 4. Make sure that **main** is selected for **base:** and **fix-sidebar** for **compare:**.
```
Looks like the original version was inconsistent with formatting. Sometimes bold, sometimes quotes, sometimes ticks. Feel free to use an existing exercise to make it consistent (or just ask me to do it since I already know what we do most of the time).

chriswblake commented on README.md:
```diff
@@ -1,24 +1,7 @@
-
 # Connect the dots in a GitHub repository
 
 _Useful tips when navigating through your repository._
```
You might consider changing this. The original description essentially says nothing.

chriswblake reviewed (changes requested):
This is a good first draft of the migration. I made some simple cleanup type changes across the files that you might not have thought to look at.

See my in-context comments for other feedback.

Finally, the main piece of feedback is to do a test run as the learner. I don't see any major changes to the contnet so it is likely very out of date. The easiest check is to do a trial run and assume you know very little. Walk through it and read it all word for word. Verify the pictures, button callouts, etc. are still accurate with github.com.

FidelusAleksander commented on .github/workflows/0-start-exercise.yml:
```diff
@@ -0,0 +1,157 @@
+          
+      - name: Prepare files history
+        run: |
+          echo "Make sure we are on step 0"
+          if [ "${{ github.workflow }}" != "Step 0" ]
+          then
+            echo "Current step is not Step 0"
+            exit 0
+          fi
```
This if statement is not needed.

FidelusAleksander commented on .github/workflows/0-start-exercise.yml:
```diff
@@ -0,0 +1,157 @@
+  prep_repo_content: 
+    name: Prep repo for next step
+    runs-on: ubuntu-latest
+    needs: [start_exercise, create_first_task]
```
I don't see a dependency that this job needs to be run after `create_first_task`, or am  I wrong?
```suggestion
    needs: [start_exercise]
```

FidelusAleksander commented on .github/workflows/0-start-exercise.yml:
```diff
@@ -0,0 +1,157 @@
+
+      - name: Build comment - add step content
+        id: build-comment
+        uses: skills/action-text-variables@v1
+        with:
+          template-file: ${{ env.STEP_1_FILE }}
+          template-vars: |
+            full_repo_name=${{ github.repository }}
+            original_issue=${{ needs.create_first_task.outputs.original_issue }}
+            duplicate_issue=${{ needs.create_first_task.outputs.duplicate_issue }}
```
Let's use the newest version of skills/action-text-variables which supports yaml formatting
```suggestion
        uses: skills/action-text-variables@v2
        with:
          template-file: ${{ env.STEP_1_FILE }}
          template-vars: |
            full_repo_name: ${{ github.repository }}
            original_issue: ${{ needs.create_first_task.outputs.original_issue }}
            duplicate_issue: ${{ needs.create_first_task.outputs.duplicate_issue }}
```

FidelusAleksander commented on .github/workflows/2-find-commit-in-history.yml:
```diff
@@ -1,81 +1,102 @@
+    name: Check Step Work
+    runs-on: ubuntu-latest
+    needs: [find_exercise, validate_commit_reference]
+    if: contains(github.event.comment.body, ${{ needs.validate_commit_reference.outputs.commit_id }})
```
We usually let the job run for wider conditions/without filters and check for what needed to be done within the job.

This gives you the option to provide feedback to the user, unlike here, they will not receive any feedback or progress in the exercise if they don't provide the commit_id, or in other cases `Duplicate of` or `Closes` comments.

Please take a look at https://github.com/skills/exercise-template/blob/main/.github/workflows/2-step.yml workflow's `check_step_work` and how it guides the user. 

I'm free to sync up and talk on this

wolz-CODElife committed:
feat: migrate exercise to issue-based flow

wolz-CODElife committed:
refactor: streamline workflow dependencies and enhance validation checks

wolz-CODElife committed:
refactor: improve pull request validation checks and streamline workflow logic

feat: add step to create comment with updated issue content

fix: remove unnecessary GH_TOKEN variable from comment creation step

fix: update comment creation step to use ISSUE_BODY variable for improved clarity

fix: correct syntax for template-vars in comment creation step

fix: streamline PR validation and disable workflow step

fix: update response templates step and improve branch name check

fix: enhance comment updates and error handling in workflow steps

chriswblake commented:
> [!NOTE]
> Publishing of these updates is delayed for business reasons. Current expectation is late July.

If you would like to use the updated version, you are welcome to take the exercise from the unpublished version (skills-dev org). Just click on the source branch `skills-dev:main`. We consider this update complete so nothing should change before launch.

wolz-CODElife committed:
fix: remove conditional checks for comment updates and streamline workflow steps

fix: update commit validation steps and streamline PR checks

fix: enhance validation steps for issue comments and streamline workflow dependencies

wolz-CODElife committed:
fix: streamline validation steps and enhance error handling in workflow checks

fix: correct indentation for check_step_work job in workflow files

fix: reorder checks for PR body and merge state in sidebar workflow

fix: enhance failure condition to include PR merge state in workflow

wolz-CODElife committed:
Merge pull request #2 from wolz-CODElife/main

feat: Migrate exercise to issue-based flow

chriswblake commented:
Migration is on pause until further notice.
